bluefog/common/nccl_controller.cc

```diff
@@ -57,23 +61,107 @@ void NCCLContext::Initialize(const int rank, const int size,
       MPI_Bcast((void*)&nccl_id, sizeof(nccl_id), MPI_BYTE, 0, MPI_COMM_WORLD));
 
   // Assume one device per process
-  CUDACHECK(cudaSetDevice(local_rank));
+  int nDevices = 0;
+  CUDACHECK(cudaGetDeviceCount(&nDevices));
+  CUDACHECK(cudaSetDevice(local_rank % nDevices));
   CUDACHECK(cudaStreamCreate(&stream));
-  ncclCommInitRank(&nccl_comm, size, nccl_id, rank);
+  NCCLCHECK(ncclCommInitRank(&nccl_comm, size, nccl_id, rank));
 
   is_initialized = true;
-  cuda_device = local_rank;
+  cuda_device = local_rank % nDevices;
   return;
 }
 
+void NCCLContext::CleanPeerCommunicator() {
```

Comment:
Add NCCL_MINOR Check to wrap these Peer Function to ensure they are not called when NCCL 2.7.3 is presented

Reply:
Done.